File: Vaidehi Gupta/ARRAY/31.next permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int i=nums.size()-2;
        while(i>=0 && (nums[i]>=nums[i+1])){
            --i;
        }  
        if(i==-1){
            reverse(nums.begin(),nums.end());
        }
        else{
            int j=nums.size()-1;
            while(j>i){
                if(nums[j]>nums[i]){
                    swap(nums[i],nums[j]);
                    break;
                }
                j--;
            }
            reverse(nums.begin()+i+1,nums.end());
        }
    }
};

Time complexity: O(n)
space complexity: O(1)

File: Vishakha Mishra /ARRAY/31. Next Permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        if (n <= 1) {
            return;
        }
        
        int idx1 = -1;
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                idx1 = i;
                break;
            }
        }
        
        if (idx1 < 0) {
            reverse(nums, 0, n - 1);
    
        } else {
            int idx2 = -1;
            for (int i = n - 1; i > idx1; i--) {
                if (nums[i] > nums[idx1]) {
                    idx2 = i;
                    break;
                }
            }
            swap(nums, idx1, idx2);
            reverse(nums, idx1 + 1, n - 1);
        }
    }
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }
}

